README.md
# Active Merchant

Active Merchant is an extraction from the e-commerce system [Shopify](http://www.shopify.com).
Shopify's requirements for a simple and unified API to access dozens of different payment
gateways with very different internal APIs was the chief principle in designing the library.

It was developed for usage in Ruby on Rails web applications and integrates seamlessly
as a Rails plugin, but it also works excellently as a stand alone Ruby library.

Active Merchant has been in production use since June 2006 and is now used in most modern
Ruby applications which deal with financial transactions. It is maintained by the
[Shopify](http://www.shopify.com) and [Spreedly](https://spreedly.com) teams, with much help
from an ever-growing set of contributors.

See {file:GettingStarted.md} if you want to learn more about using Active Merchant in your
applications.

## Installation

### From Git

You can check out the latest source from git:

    git clone git://github.com/Shopify/active_merchant.git

### From RubyGems

Installation from RubyGems:

    gem install activemerchant

Or, if you're using Bundler, just add the following to your Gemfile:

    gem 'activemerchant'

## Usage

This simple example demonstrates how a purchase can be made using a person's
credit card details.

    require 'rubygems'
    require 'active_merchant'

    # Use the TrustCommerce test servers
    ActiveMerchant::Billing::Base.mode = :test

    gateway = ActiveMerchant::Billing::TrustCommerceGateway.new(
                :login => 'TestMerchant',
                :password => 'password')

    # ActiveMerchant accepts all amounts as Integer values in cents
    amount = 1000  # $10.00

    # The card verification value is also known as CVV2, CVC2, or CID
    credit_card = ActiveMerchant::Billing::CreditCard.new(
                    :first_name         => 'Bob',
                    :last_name          => 'Bobsen',
                    :number             => '4242424242424242',
                    :month              => '8',
                    :year               => Time.now.year+1,
                    :verification_value => '000')

    # Validating the card automatically detects the card type
    if credit_card.valid?
      # Capture $10 from the credit card
      response = gateway.purchase(amount, credit_card)

      if response.success?
        puts "Successfully charged $#{sprintf("%.2f", amount / 100)} to the credit card #{credit_card.display_number}"
      else
        raise StandardError, response.message
      end
    end

For more in-depth documentation and tutorials, see {file:GettingStarted.md} and the
[API documentation](http://rubydoc.info/github/Shopify/active_merchant/master/file/README.md).

## Supported Direct Payment Gateways

The [ActiveMerchant Wiki](http://github.com/Shopify/active_merchant/wikis) contains a [table of features supported by each gateway](http://github.com/Shopify/active_merchant/wikis/gatewayfeaturematrix).

* [Authorize.Net](http://www.authorize.net/) - US
* [Authorize.Net CIM](http://www.authorize.net/) - US
* [Balanced](https://www.balancedpayments.com/) - US
* [Banwire](https://www.banwire.com/) - MX
* [Barclays ePDQ](http://www.barclaycard.co.uk/business/accepting-payments/epdq-mpi/) - UK
* [Be2Bill](http://www.be2bill.com/) - FR
* [Beanstream.com](http://www.beanstream.com/) - CA
* [BluePay](http://www.bluepay.com/) - US
* [Braintree](http://www.braintreepaymentsolutions.com) - US
* [CardStream](http://www.cardstream.com/) - UK
* [CertoDirect](http://www.certodirect.com/) - BE, BG, CZ, DK, DE, EE, IE, EL, ES, FR, IT, CY, LV, LT, LU, HU, MT, NL, AT, PL, PT, RO, SI, SK, FI, SE, UK
* [CyberSource](http://www.cybersource.com) - US
* [DataCash](http://www.datacash.com/) - UK
* [Efsnet](http://www.concordefsnet.com/) - US
* [Elavon MyVirtualMerchant](http://www.elavon.com) - US, CA
* [ePay](http://www.epay.dk/) - DK, SE, NO
* [eWAY](http://www.eway.com.au/) - AU
* [eWay Rapid 3.0](http://www.eway.com.au/) - AU
* [E-xact](http://www.e-xact.com) - CA, US
* [Fat Zebra](https://www.fatzebra.com.au) - AU
* [Federated Canada](http://www.federatedcanada.com/) - CA
* [FirstData Global Gateway e4](http://www.firstdata.com) - CA, US
* [FirstPay](http://www.first-pay.com) - US
* [Garanti Sanal POS](https://ccpos.garanti.com.tr/ccRaporlar/garanti/ccReports) - US, TR
* [HDFC](http://www.hdfcbank.com/sme/sme-details/merchant-services/guzh6m0i) - IN
* [Inspire](http://www.inspiregateway.com) - US
* [InstaPay](http://www.instapayllc.com) - US
* [Iridium](http://www.iridiumcorp.co.uk/) - UK, ES
* [iTransact](http://www.itransact.com/) - US
* [JetPay](http://www.jetpay.com) - US
* [LinkPoint](http://www.linkpoint.com/) - US
* [Litle](http://www.litle.com/) - US
* [Merchant e-Solutions](http://merchante-solutions.com/) - US
* [MerchantWare](http://merchantwarehouse.com/merchantware) - US
* [Mercury](http://www.mercurypay.com) - US
* [MasterCard Internet Gateway Service (MiGS)](http://mastercard.com/mastercardsps) - AU, AE, BD, BN, EG, HK, ID, IN, JO, KW, LB, LK, MU, MV, MY, NZ, OM, PH, QA, SA, SG, TT, VN
* [Modern Payments](http://www.modpay.com) - US
* [Moneris](http://www.moneris.com/) - CA
* [Moneris US](http://www.monerisusa.com/) - US
* [NABTransact](http://www.nab.com.au/nabtransact/) - AU
* [NELiX TransaX Gateway](http://www.nelixtransax.com) - US
* [Netaxept](http://www.betalingsterminal.no/Netthandel-forside) - NO, DK, SE, FI
* [NETbilling](http://www.netbilling.com) - US
* [NetPay](http://www.netpay.com.mx) - MX
* [NetRegistry](http://www.netregistry.com.au) - AU
* [NMI](http://nmi.com/) - US
* [Ogone DirectLink](http://www.ogone.com) - BE, DE, FR, NL, AT, CH
* [Optimal Payments](http://www.optimalpayments.com/) - CA, US, UK
* [Orbital Paymentech](http://chasepaymentech.com/) - CA, US, UK, GB
* [PayBox Direct](http://www.paybox.com) - FR
* [PayGate PayXML](http://paygate.co.za/) - US, ZA
* [PayJunction](http://www.payjunction.com/) - US
* [PaymentExpress](http://www.paymentexpress.com/) - AU, MY, NZ, SG, ZA, UK, US
* [PayPal Express Checkout](https://www.paypal.com/cgi-bin/webscr?cmd=xpt/merchant/ExpressCheckoutIntro-outside) - US, CA, SG, AU
* [PayPal Payflow Pro](https://www.paypal.com/cgi-bin/webscr?cmd=_payflow-pro-overview-outside) - US, CA, SG, AU
* [PayPal Website Payments Pro (UK)](https://www.paypal.com/uk/cgi-bin/webscr?cmd=_wp-pro-overview-outside) - UK
* [PayPal Website Payments Pro (CA)](https://www.paypal.com/cgi-bin/webscr?cmd=_wp-pro-overview-outside) - CA
* [PayPal Express Checkout](https://www.paypal.com/cgi-bin/webscr?cmd=xpt/merchant/ExpressCheckoutIntro-outside) - US
* [PayPal Website Payments Pro (US)](https://www.paypal.com/cgi-bin/webscr?cmd=_wp-pro-overview-outside) - US
* [PaySecure](http://www.commsecure.com.au/paysecure.shtml) - AU
* [PayWay](https://www.payway.com.au) - AU
* [Plug'n Pay](http://www.plugnpay.com/) - US
* [Psigate](http://www.psigate.com/) - CA
* [PSL Payment Solutions](http://www.paymentsolutionsltd.com/) - UK
* [Quantum](http://www.quantumgateway.com) - US
* [QuickBooks Merchant Services](http://payments.intuit.com/) - US
* [Quickpay](http://quickpay.dk/) - DK, SE
* [Rabobank Nederland](http://www.rabobank.nl/) - NL
* [Realex](http://www.realexpayments.com/) - IE, UK
* [Redsys](http://www.redsys.es) - ES
* [SagePay](http://www.sagepay.com) - UK
* [Sage Payment Solutions](http://www.sagepayments.com) - US, CA
* [Sallie Mae](http://www.salliemae.com) - US
* [SecureNet](http://www.securenet.com) - US
* [SecurePay](http://securepay.com.au) - AU
* [SecurePay](http://www.securepay.com/) - US
* [SecurePayTech](http://www.securepaytech.com/) - NZ
* [SkipJack](http://www.skipjack.com/) - US, CA
* [Spreedly Core](https://spreedlycore.com/) - AD, AE, AT, AU, BD, BE, BG, BN, CA, CH, CY, CZ, DE, DK, EE, EG, ES, FI, FR, GB, GI, GR, HK, HU, ID, IE, IL, IM, IN, IS, IT, JO, KW, LB, LI, LK, LT, LU, LV, MC, MT, MU, MV, MX, MY, NL, NO, NZ, OM, PH, PL, PT, QA, RO, SA, SE, SG, SI, SK, SM, TR, TT, UM, US, VA, VN, ZA
* [Stripe](https://stripe.com/) - US
* [TransFirst](http://www.transfirst.com/) - US
* [TrustCommerce](http://www.trustcommerce.com/) - US
* [USA ePay](http://www.usaepay.com/) - US
* [Verifi](http://www.verifi.com/) - US
* [ViaKLIX](http://viaklix.com) - US
* [Vindica](http://www.vindicia.com/) - US, CA, UK, AU, MX, BR, DE, KR, CN, HK
* [WebPay](https://webpay.jp/) - JP
* [Wirecard](http://www.wirecard.com) - DE
* [WorldPay](http://www.worldpay.com) - AU, HK, UK, US

## Supported Offsite Payment Gateways

* [2 Checkout](http://www.2checkout.com)
* [A1Agregator](http://a1agregator.ru/) - RU
* [Authorize.Net SIM](http://developer.authorize.net/api/sim/) - US
* [Banca Sella GestPay](https://www.sella.it/banca/ecommerce/gestpay/gestpay.jsp)
* [Chronopay](http://www.chronopay.com)
* [DirecPay](http://www.timesofmoney.com/direcpay/jsp/home.jsp)
* [Direct-eBanking / sofortueberweisung.de by Payment-Networks AG](https://www.payment-network.com/deb_com_en/merchantarea/home) - DE, AT, CH, BE, UK, NL
* [Dotpay](http://dotpay.pl)
* [Dwolla](https://www.dwolla.com/default.aspx)
* [ePay](http://www.epay.dk/epay-payment-solutions/)
* [First Data](https://firstdata.zendesk.com/entries/407522-first-data-global-gateway-e4sm-payment-pages-integration-manual)
* [HiTRUST](http://www.hitrust.com.hk/)
* [Moneybookers](http://www.moneybookers.com)
* [Nochex](http://www.nochex.com)
* [Paxum](https://www.paxum.com/)
* [PayPal Website Payments Standard](https://www.paypal.com/cgi-bin/webscr?cmd#_wp-standard-overview-outside)
* [Paysbuy](https://www.paysbuy.com/) - TH
* [Robokassa](http://robokassa.ru/)
* [SagePay Form](http://www.sagepay.com/products_services/sage_pay_go/integration/form)
* [Suomen Maksuturva](https://www.maksuturva.fi/services/vendor_services/integration_guidelines.html)
* [Valitor](http://www.valitor.is/) - IS
* [Verkkomaksut](http://www.verkkomaksut.fi) - FI
* [WebMoney](http://www.webmoney.ru) - RU
* [WebPay](http://webpay.by/)
* [WorldPay](http://www.worldpay.com)

## Contributing

The source code is hosted at [GitHub](http://github.com/Shopify/active_merchant), and can be fetched using:

    git clone git://github.com/Shopify/active_merchant.git

Please see the [ActiveMerchant Guide to Contributing](http://github.com/Shopify/active_merchant/wikis/contributing) for
information on adding a new gateway to ActiveMerchant.

Please don't touch the CHANGELOG in your pull requests, we'll add the appropriate CHANGELOG entries
at release time.

[![Build Status](https://secure.travis-ci.org/Shopify/active_merchant.png)](http://travis-ci.org/Shopify/active_merchant)

[![Code Climate](https://codeclimate.com/badge.png)](https://codeclimate.com/github/Shopify/active_merchant)


CHANGELOG
= ActiveMerchant CHANGELOG

* Add FirstData Global Gateway e4 [frobcode]
* PaymentExpress: Add support for optional fields: ClientType and TxnData [moklett]
* PaymentExpress: Limit MerchantReference/description to 64 chars [moklett]
* Wirecard: description must be no more than 32 characters [moklett]
* Litle: Add support for passing a token to the authorize and purchase methods [forest]
* PayPal Common: Allow searching for transactions by ProfileID [aq1018]
* Add Spreedly Core gateway [duff]
* eWay Gateway: Return proper value for authorization [duff]
* eWay Gateway: Add support for refunds [duff]
* Quickpay: Add support for protocols 5 & 6 [twarberg]
* Banwire gateway: Handle JSON::ParserError [duff]
* Balanced gateway: Fix unspecified marketplace [duff]
* QBMS gateway: Allow partial addresses [duff]
* Authorize.Net CIM: Allow omitting card expiration date [shanebonham]
* Authorize.Net CIM: Add support for extraOptions to createCustomerProfileTransaction [tpiekos]
* Add NETPAY gateway [samlown]
* Balanced gateway: Add amount to the refund method signature [ntalbott]
* Orbital gateway: Fix void method signature [aprofeit, ntalbott]
* Eway Managed: Add 'query_customer' API as #retrieve [cdaloisio]
* NetPay: Fix the signature for void [duff]
* Cybersource: Add check support [bowmande]
* Add Be2Bill gateway [michaelbaudino]
* Moneris: Use a capture of $0 for void [ntalbott]
* PayPal Express integration: Fix received_at time zone [ntalbott]
* NAB Transact: Add refund capability [nagash]
* Stripe: Add support for application_fee [duff]
* SagePay: Add support for GiftAidPayment [duff]
* Wirecard: Add support for partial captures [richardblair]

== Version 1.29.3 (December 7, 2012)

* Braintree Blue: Better wiredump_device support [ntalbott]
* Braintree: Store sets vault id as authorization [ntalbott]
* WorldPay: Fix currencies without fractions like JPY and HUF by rounding down amount [Soleone]

== Version 1.29.2 (December 7, 2012)

* Moneris: fix issue with the default options not being merged [jduff]
* Sage Pay: Make 0000 default post code for everyone if missing [BlakeMesdag]

== Version 1.29.1 (December 5, 2012)

* Add eWay Rapid 3.0 gateway [ntalbott]
* Fix AVS responses missing attributes [jduff]

== Version 1.29.0 (November 30, 2012)

* Authorize.Net gateway: Support description and order_id for capture [ntalbott]
* Add Mercury gateway [adr1anx, opendining]
* Webmoney integration: Add gross, item_id, and amount accessors to notification [fr33z3]
* Fix running tests under ActiveSupport 2.3.14 [ntalbott]
* SagePay Form integration: Remove dependency on ActiveSupport's String#truncate [ntalbott]
* Elavon gateway: Add support for the sales tax parameter [stevestmartin]
* Add WebPay gateway [keikubo]
* iTransact gateway: make void API consistent [frobcode]
* Stripe gateway: Pass city in add_address [npverni]
* Paypal gateway: Add option to change the outstanding balance of a recurring billing profile [mattwhite]
* Mercury gateway: Fix authorizations API [ntalbott]
* Mercury gateway: Support refund properly [ntalbott]
* Braintree Blue gateway: Add support for the recurring flag [ntalbott]
* Add HDFC gateway [ntalbott]
* HDFC gateway: Add more supported currencies [ntalbott]
* HDFC gateway: Add support for passing ECI value [ntalbott]
* HDFC gateway: Allow setting test mode via options [ntalbott]
* HDFC gateway: Pass phone number in UDF3 [ntalbott]
* HDFC gateway: Fix unescaped '&' entity in XML [ntalbott]
* HDFC gateway: More robust entity fixing [ntalbott]
* Add A1Agregator integration [england]
* Refactored handling of #test? and @options [ntalbott]
* Realex gateway: Realex gateway: Fix billing address format [ntalbott]
* Orbital gateway: handle custom AVS response codes [jonm-okc]
* Orbital gateway: Fix infinite connection retry [jonm-okc, ntalbott]
* PayPal integration: Add support for MassPay IPN notifications [damonmorgan]
* Orbital gateway: Fix status of void result [jonm-okc]
* Add Redsys gateway [samlown]
* Cybersource gateway: Add support for subscription credit [fabiokr]
* Use Thor for generators [ntalbott]
* Psigate gateway: Add void support [samuelreh]
* Add Liqpay integration [beorc]
* Paypal Express gateway: Add shipping accessor to response [v-fedorov]

== Version 1.28.0 (August 10, 2012)

* PayPal Express: support non standard locale codes [Soleone]
* Litle: allow setting test mode per transaction [jduff]
* Add Banwire gateway [acolin]
* Authorize.Net CIM gateway: Move cardCode after order to comply with the XSD [davetron5000]
* Add WebMoney integration [Mehonoshin]
* EasyPay: Make symmetric with other integrations [nashby]
* Add Paysbuy integration [divineforest]
* Bogus gateway: Use last digit for pass/fail [mipearson]
* Elavon gateway: Separate from Viaklix, implement refund & void [duff]
* Orbital gateway: Update to API version 5.6 and add support for profile requests [rbarazi]

== Version 1.27.0 (August 10, 2012)

* Add First Data integration [courtland]
* Add WebPay integration [nashby]
* Add Suomen Maksuturva integration [akonan]
* Payway gateway: Fix card storage [BenZhang]
* Payflow Pro gateway: Add MaxFailPayments support [gregwinn]
* Add Paxum integration [Mehonoshin]
* Add Balanced gateway [mjallday]
* Plug'n Pay gateway: Add tests for partial capture [csaunders]
* Braintree Blue gateway: Add credit card details to responses [dougbradbury]
* PayPal gateway: Support for 'Full' refund type [kurenn]
* Worldpay: fix refund [jduff/omh]
* Add PxPay offsite integration [boourns]
* Wirecard: always capture 'authorization' transaction [ntalbott]
* Add rake task to verify ssl certs [boourns]

== Version 1.26.0 (July 6, 2012)

* Orbital gateway: fix broken requests by ensuring the order of XML elements matches their DTD [Soleone]
* CyberSource gateway: clean up formatting [ntalbott]
* Netbilling gateway: Add refund/credit/void support [ntalbott]
* Add PayGate XML gateway [rubyisbeautiful]
* Add PayWay gateway [BenZhang]

test/unit/gateways/be2bill_test.rb
require 'test_helper'

class Be2billTest < Test::Unit::TestCase
  def setup
    @gateway = Be2billGateway.new(:login => 'login', :password  => 'password')

    @credit_card = credit_card
    @amount = 1000

    @options = {
      :order_id         => 1,
      :address          => address,
      :customer         => 'Guest #42',
      :description      => 'A ski pass from SkiWallet',
      :client_useragent => 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Ubuntu Chromium/23.0.1271.97 Chrome/23.0.1271.97 Safari/537.11',
      :client_ip        => '42.42.42.42',
      :client_email     => 'georges.abitbol@alpine-lab.com'
    }
  end

  def test_supported_countries
    assert_equal ['FR'], Be2billGateway.supported_countries
  end

  def test_supported_card_types
    assert_equal [:visa, :master, :american_express], Be2billGateway.supported_cardtypes
  end

  def test_add_amount
    params = {}
    @gateway.send(:add_amount, params, 1000)
    assert_equal params, {:AMOUNT => 1000}
  end

  def test_add_order
    params = {}
    @gateway.send(:add_order, params, :order_id => 1)
    assert_equal params, {:ORDERID => 1}
  end

  def test_add_customer
    params = {}
    options = {
      :customer         => 'Georges Abitbol',
      :client_useragent => 'w3m/ftw',
      :client_ip        => '10.73.73.73',
      :client_email     => 'georges.abitbol@alpine-lab.com'
    }
    @gateway.send(:add_customer, params, options)
    assert_equal params, {
      :CLIENTIDENT     => 'Georges Abitbol',
      :CLIENTUSERAGENT => 'w3m/ftw',
      :CLIENTIP        => '10.73.73.73',
      :CLIENTEMAIL     => 'georges.abitbol@alpine-lab.com'
    }
  end

  def test_add_description
    params = {}
    @gateway.send(:add_description, params, :description => 'A simple description')
    assert_equal params, {:DESCRIPTION => 'A simple description'}
  end

  def test_add_creditcard
    params = {}
    cc = CreditCard.new({
      :number             => '4111111111111111',
      :month              => 9,
      :year               => 2017,
      :first_name         => 'Georges',
      :last_name          => 'Abitbol',
      :verification_value => '123',
      :brand              => 'visa'
    })
    @gateway.send(:add_creditcard, params, cc)
    assert_equal params, {
      :CARDCODE         => '4111111111111111',
      :CARDFULLNAME     => 'Georges Abitbol',
      :CARDVALIDITYDATE => '09-17',
      :CARDCVV          => '123'
    }
  end

  def test_add_3dsecure
    # Valid 3DSecure options
    params = {}
    @gateway.send(:add_3dsecure, params, :'3dsecure' => true, :'3dsecuremode' => 'POPUP')
    assert_equal params, {:'3DSECURE' => 'yes', :'3DSECUREDISPLAYMODE' => 'POPUP'}
    # Another valid 3DSecure options
    params = {}
    @gateway.send(:add_3dsecure, params, :'3dsecure' => 'yes', :'3dsecuremode' => 'popup')
    assert_equal params, {:'3DSECURE' => 'yes', :'3DSECUREDISPLAYMODE' => 'POPUP'}
    # Invalid 3DSecure options
    params = {}
    @gateway.send(:add_3dsecure, params, :'3dsecure' => 'Georges', :'3dsecuremode' => 'Abitbol')
    assert_equal params, {}
  end

  def test_be2bill_digest
    params = {
      :key1 => :value1,
      :key2 => :value2,
      :key3 => :value3
    }
    response = @gateway.send(:be2bill_digest, params)
    assert_equal response, '154f6aec5f22fc6fc1c85a6968a2f6f7e614e666b40b09acbcbfa965fe1c5f9e'
  end

  def test_post_data
    params = {
      :key1 => :value1,
      :key2 => :value2,
      :key3 => :value3
    }
    response = @gateway.send(:post_data, 'payment', params)
    assert_equal response, 'method=payment&params%5Bkey1%5D=value1&params%5Bkey2%5D=value2&params%5Bkey3%5D=value3'
  end

  def test_response_is_valid
    # Valid successful response
    assert @gateway.send(:response_is_valid?, {
      'EXECCODE'      => '0000',
      'OPERATIONTYPE' => 'payment',
      'MESSAGE'       => 'Message a caractere informatif',
      'TRANSACTIONID' => 'AB12345678'
    })
    # Valid failure response
    assert @gateway.send(:response_is_valid?, {
      'EXECCODE'      => '4001',
      'OPERATIONTYPE' => 'payment',
      'MESSAGE'       => 'Message a caractere informatif'
    })
    # Invalid response (lacks EXECCODE)
    assert !@gateway.send(:response_is_valid?, {
      'OPERATIONTYPE' => 'payment',
      'MESSAGE'       => 'Message a caractere informatif',
      'TRANSACTIONID' => 'AB12345678'
    })
    # Invalid response (lacks OPERATIONTYPE)
    assert !@gateway.send(:response_is_valid?, {
      'EXECCODE'      => '0000',
      'MESSAGE'       => 'Message a caractere informatif',
      'TRANSACTIONID' => 'AB12345678'
    })
    # Invalid response (lacks MESSAGE)
    assert !@gateway.send(:response_is_valid?, {
      'EXECCODE'      => '0000',
      'OPERATIONTYPE' => 'payment',
      'TRANSACTIONID' => 'AB12345678'
    })
    # Invalid response (lacks TRANSACTIONID while EXECCODE indicates success)
    assert !@gateway.send(:response_is_valid?, {
      'EXECCODE'      => '0000',
      'MESSAGE'       => 'Message a caractere informatif',
      'OPERATIONTYPE' => 'payment'
    })
    # Invalid response (invalid EXECCODE)
    assert !@gateway.send(:response_is_valid?, {
      'EXECCODE'      => '000',
      'OPERATIONTYPE' => 'payment',
      'MESSAGE'       => 'Message a caractere informatif',
      'TRANSACTIONID' => 'AB12345678'
    })
    # Invalid response (invalid OPERATIONTYPE)
    assert !@gateway.send(:response_is_valid?, {
      'EXECCODE'      => '0000',
      'OPERATIONTYPE' => 'not_an_action',
      'MESSAGE'       => 'Message a caractere informatif',
      'TRANSACTIONID' => 'AB12345678'
    })
  end

  def test_response_is_success
    # Successful response
    assert @gateway.send(:response_is_success?, {
      'EXECCODE'      => '0000'
    })
    # Failure response
    assert !@gateway.send(:response_is_success?, {
      'EXECCODE'      => '4001'
    })
  end

  def test_be2bill_response_class
    # Response does require 3DSecure
    @gateway.expects(:ssl_post).returns(purchase_requires_3dsecure_response)
    response = @gateway.purchase(@amount, @credit_card, @options)
    assert_instance_of ActiveMerchant::Billing::Be2billResponse, response
    assert response.respond_to? :requires_3dsecure?
    assert response.send(:requires_3dsecure?)
    # Response does *not* require 3DSecure
    @gateway.expects(:ssl_post).returns(successful_purchase_response)
    response = @gateway.purchase(@amount, @credit_card, @options)
    assert_instance_of ActiveMerchant::Billing::Be2billResponse, response
    assert response.respond_to? :requires_3dsecure?
    assert !response.send(:requires_3dsecure?)
  end

  def test_successful_purchase
    @gateway.expects(:ssl_post).returns(successful_purchase_response)
    assert response = @gateway.purchase(@amount, @credit_card, @options)

    assert_instance_of ActiveMerchant::Billing::Be2billResponse, response
    assert_success response
    assert response.test?
    assert_equal '0000', response.params['EXECCODE']
    assert_equal response.authorization, 'AB12345678'
  end

  def test_unsuccessful_purchase
    @gateway.expects(:ssl_post).returns(failed_purchase_response)
    assert response = @gateway.purchase(@amount, @credit_card, @options)

    assert_instance_of ActiveMerchant::Billing::Be2billResponse, response
    assert_failure response
    assert response.test?
    assert_match response.params['EXECCODE'], /^[1-9][0-9]{3}$/
  end

  private

  def successful_purchase_response
    '{"OPERATIONTYPE":"payment","TRANSACTIONID":"AB12345678","EXECCODE":"0000","MESSAGE":"The transaction has been accepted.","DESCRIPTOR":"Skiwallet"}'
  end

  def purchase_requires_3dsecure_response
    '{"OPERATIONTYPE":"payment","TRANSACTIONID":"AB12345678","EXECCODE":"0001","MESSAGE":"The transaction requires 3DSecure authentification.","DESCRIPTOR":"Skiwallet"}'
  end

  def failed_purchase_response
    '{"OPERATIONTYPE":"payment","TRANSACTIONID":"AB12345678","EXECCODE":"4001","MESSAGE":"The bank refused the transaction."}'
  end

end


test/remote/gateways/remote_be2bill_test.rb
# encoding: utf-8
require 'test_helper'

class RemoteBe2billTest < Test::Unit::TestCase

  def setup
    @gateway = Be2billGateway.new(fixtures(:be2bill))

    @amount = 1000
    @credit_card = credit_card('5555556778250000')
    @declined_card = credit_card('5555557376384001')

    @options = {
      :order_id         => 1,
      :address          => address,
      :customer         => 'Guest #42',
      :description      => 'A ski pass from SkiWallet',
      :client_useragent => 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.11 (KHTML, like Gecko) Ubuntu Chromium/23.0.1271.97 Chrome/23.0.1271.97 Safari/537.11',
      :client_ip        => '42.42.42.42',
      :client_email     => 'georges.abitbol@alpine-lab.com'
    }
  end

  def test_successful_purchase
    assert response = @gateway.purchase(@amount, @credit_card, @options)
    assert_success response
    assert_equal response.params['EXECCODE'], '0000'
  end

  def test_successful_with_utf8_encoding
    assert response = @gateway.purchase(@amount, @credit_card, @options.merge(:customer => 'Pétère et Stévène'))
    assert_success response
    assert_equal response.params['EXECCODE'], '0000'
  end

  def test_unsuccessful_purchase
    assert response = @gateway.purchase(@amount, @declined_card, @options)
    assert_failure response
    assert_equal response.params['EXECCODE'], '4001'
  end

  # def test_authorize_and_capture
  #   amount = @amount
  #   assert auth = @gateway.authorize(amount, @credit_card, @options)
  #   assert_success auth
  #   assert_equal 'Success', auth.message
  #   assert auth.authorization
  #   assert capture = @gateway.capture(amount, auth.authorization)
  #   assert_success capture
  # end

  # def test_failed_capture
  #   assert response = @gateway.capture(@amount, '')
  #   assert_failure response
  #   assert_equal 'REPLACE WITH GATEWAY FAILURE MESSAGE', response.message
  # end

  def test_invalid_login
    gateway = Be2billGateway.new(:login => '', :password => '')
    assert response = gateway.purchase(@amount, @credit_card, @options)
    assert_failure response
    assert_equal response.params['EXECCODE'], '1001'
  end
end


lib/active_merchant/billing/gateways/be2bill.rb
module ActiveMerchant #:nodoc:
  module Billing #:nodoc:
    class Be2billGateway < Gateway

      # Gateway global variables
      self.display_name = 'Be2Bill'
      self.homepage_url = 'http://www.be2bill.com/'
      self.supported_countries = ['FR']
      self.supported_cardtypes = [:visa, :master, :american_express]
      self.default_currency = :EUR
      self.money_format = :cents
      self.test_url = 'https://secure-test.be2bill.com/front/service/rest/process.php'
      self.live_url = 'https://secure-magenta1.be2bill.com/front/service/rest/process.php'
      API_VERSION = '2.0'

      def initialize(options = {})
        requires!(options, :login, :password)
        super
      end

      def purchase(money, creditcard, options = {})
        params = {}
        add_amount(params, money)
        add_order(params, options)
        add_customer(params, options)
        add_description(params, options)
        add_creditcard(params, creditcard)
        add_3dsecure(params, options)
        commit(:payment, params)
      end

      def authorize(money, creditcard, options = {})
        # TODO
        # commit('authorize', ...)
      end

      def capture(money, authorization, options = {})
        # TODO
        # commit('capture', ...)
      end

      def refund(money, reference, options = {})
        # TODO
        # commit('refund', ...)
      end

      def credit(money, identification_or_credit_card, options = {})
        # TODO
        # commit('credit', ...)
      end

      private

      def add_amount(params, amount)
        params[:AMOUNT] = amount
      end

      def add_order(params, options)
        params[:ORDERID] = options[:order_id]
      end

      def add_customer(params, options)
        params[:CLIENTIDENT] = options[:customer]
        params[:CLIENTUSERAGENT] = options[:client_useragent]
        params[:CLIENTIP] = options[:client_ip]
        params[:CLIENTEMAIL] = options[:client_email]
      end

      def add_description(params, options)
        params[:DESCRIPTION] = options[:description]
      end

      def add_creditcard(params, creditcard)
        params[:CARDCODE] = creditcard.number
        params[:CARDFULLNAME] = "#{creditcard.first_name} #{creditcard.last_name}"
        params[:CARDVALIDITYDATE] = "#{"%02d" % creditcard.month}-#{"%02d" % (creditcard.year % 100)}"
        params[:CARDCVV] = creditcard.verification_value
      end

      def add_3dsecure(params, options)
        if options[:'3dsecure'].is_a? String and ['yes', 'no'].include? options[:'3dsecure'].downcase
          params[:'3DSECURE'] = options[:'3dsecure'].downcase
        elsif [true, false].include? options[:'3dsecure']
          params[:'3DSECURE'] = (options[:'3dsecure'] ? 'yes' : 'no')
        end
        if options[:'3dsecuremode'].is_a? String and ['MAIN', 'POPUP', 'TOP'].include? options[:'3dsecuremode'].upcase
          params[:'3DSECUREDISPLAYMODE'] = options[:'3dsecuremode'].upcase
        end
      end

      def be2bill_digest(params = {})
        clear_str = "#{@options[:password]}"
        params.sort.each do |key, value|
          clear_str << "#{key}=#{value}#{@options[:password]}"
        end
        Digest::SHA2.new(256) << clear_str
      end

      def post_data(action, params)
        {method: action, params: params}.to_query
      end

      def commit(action, params, options = {})
        params[:OPERATIONTYPE] = action
        params[:IDENTIFIER] = @options[:login]
        params[:VERSION] = API_VERSION
        params[:HASH] = be2bill_digest(params)
        url = test? ? self.test_url : self.live_url
        json_response = ssl_post(url, post_data(action, params))
        response = JSON.parse(json_response)
        if response_is_valid? response
          Be2billResponse.new(response_is_success?(response), response['MESSAGE'], response, {
          :test          => ActiveMerchant::Billing::Base.mode == :test,
          :authorization => response['TRANSACTIONID']
          })
        else
          Be2billResponse.new(false, 'Unprocessable response.')
        end
      end

      def response_is_valid?(response)
        response['EXECCODE'] =~ /[0-9]{4}/ &&
        ['payment', 'authorization', 'capture', 'refund', 'credit'].include?(response['OPERATIONTYPE']) &&
        response.has_key?('MESSAGE') &&
        (response['EXECCODE'][0] != '0' || response.has_key?('TRANSACTIONID'))
      end

      def response_is_success?(response)
        response['EXECCODE'] == '0000'
      end

    end

    class Be2billResponse < Response
      def requires_3dsecure?
        @params['EXECCODE'] == '0001'
      end
    end

  end
end

